ibis/backends/duckdb/__init__.py

```diff
@@ -208,6 +209,55 @@ def _register_failure(self):
             f"please call one of {msg} directly"
         )
 
+    def read_json(
+        self,
+        source_list: str | list[str] | tuple[str],
+        table_name: str | None = None,
+        **kwargs,
+    ) -> ir.Table:
+        """Read newline-delimited JSON into an ibis table.
+
+        Parameters
+        ----------
+        source_list
+            File or list of files
+        table_name
+            Optional table name
+        kwargs
+            Additional keyword arguments passed to DuckDB's `read_json_objects` function
+
+        Returns
+        -------
+        Table
+            An ibis table expression
+        """
+        if not table_name:
+            table_name = f"ibis_read_json_{next(json_n)}"
+
+        source_list = normalize_filenames(source_list)
+
+        objects = (
+            sa.func.read_json_objects(
+                sa.func.list_value(*source_list), _format_kwargs(kwargs)
+            )
+            .table_valued("raw")
+            .render_derived()
+        )
+        # read a single row out to get the schema, assumes the first row is representative
+        json_structure_query = sa.select(sa.func.json_structure(objects.c.raw)).limit(1)
```

Comment:
I tried this with the following JSON file and got an error from ibis

```
{"x": 1, "y": 2, "z": null}
{"x": 3, "y": 4, "z": 5}
```

Results in

```
>>> ibis.read_json("test.json")
...
ParseError: expected one of 'bigint', 'binary', 'blob', 'bool', 'boolean', 'bpchar', 'bytea', 'char', 'date', 'datetime', 'decimal', 'double', 'float', 'float4', 'float8', 'int', 'int1', 'int2', 'int4', 'int8', 'integer', 'interval', 'json', 'list', 'logical', 'long', 'map', 'numeric', 'real', 'short', 'signed', 'smallint', 'string', 'struct', 'text', 'time', 'timestamp with time zone', 'timestamp', 'timestamp_ms', 'timestamp_ns', 'timestamp_sec', 'timestamp_tz', 'tinyint', 'ubigint', 'uinteger', 'usmallint', 'utinyint', 'uuid', 'varbinary', 'varchar' at 0:0
```

The issue seems to be parsing `NULL` as a datatype doesn't work. However, fixing that results in silently dropping of column data, since duckdb will happily cast non-null data to `NULL`:

```
D select json_transform(json, '{"x": "UBIGINT", "y": "UBIGINT", "z": "NULL"}') FROM read_json_objects("test.json");
┌─────────────────────────────────────────────────────────────────────────┐
│ json_transform("json", '{"x": "UBIGINT", "y": "UBIGINT", "z": "NULL"}') │
│                 struct(x ubigint, y ubigint, z integer)                 │
├─────────────────────────────────────────────────────────────────────────┤
│ {'x': 1, 'y': 2, 'z': NULL}                                             │
│ {'x': 3, 'y': 4, 'z': NULL}                                             │
└─────────────────────────────────────────────────────────────────────────┘
```

- I wonder if we should select the first non-null value from every column and use that to infer the datatype?
- Alternatively, could/should we support an optional `schema` kwarg that takes an ibis schema for the expected table? This may be nice too, since it may allow for only extracting relevant fields from the input data.

Reply:
I think we probably want to do the latter, since the former has weird and annoying edge case behavior (all nulls except the last value).

Reply:
I have some test data I'd like to try this on, mind if I take a stab at improving the resiliency here in a follow up PR?

Reply:
Heck yeah!